libsrc/pylith/meshio/MeshIOPETSc.cc

```diff
@@ -0,0 +1,145 @@
+namespace pylith {
+    namespace meshio {
+    } // meshio
+} // pylith
+
+// ---------------------------------------------------------------------------------------------------------------------
+// Constructor
+pylith::meshio::MeshIOPETSc::MeshIOPETSc(void) :
+    _useIndexZero(true) { // constructor
+    PyreComponent::setName("meshiopetsc");
+} // constructor
+
+
+// ---------------------------------------------------------------------------------------------------------------------
+// Destructor
+pylith::meshio::MeshIOPETSc::~MeshIOPETSc(void) { // destructor
+    deallocate();
+} // destructor
+
+
+// ---------------------------------------------------------------------------------------------------------------------
+// Deallocate PETSc and local data structures.
+void
+pylith::meshio::MeshIOPETSc::deallocate(void) { // deallocate
+    PYLITH_METHOD_BEGIN;
+
+    MeshIO::deallocate();
+
+    PYLITH_METHOD_END;
+} // deallocate
+
+
+// ---------------------------------------------------------------------------------------------------------------------
+// Read mesh.
+void
+pylith::meshio::MeshIOPETSc::_read(void) { // _read
+    PYLITH_METHOD_BEGIN;
+    PYLITH_COMPONENT_DEBUG("_read()");
+    
+    assert(_mesh);
+    MPI_Comm comm = _mesh->getComm();
+    PetscErrorCode err = 0;
+    PetscDM dmMesh = NULL;
+
+    err = DMCreate(comm, &dmMesh);PYLITH_CHECK_ERROR(err);
+    err = DMSetType(dmMesh, DMPLEX);PYLITH_CHECK_ERROR(err);
+    err = DMSetFromOptions(dmMesh);PYLITH_CHECK_ERROR(err);
+    err = DMViewFromOptions(dmMesh, NULL, "-dm_view");PYLITH_CHECK_ERROR(err);
+
+    topology::Stratum cellsStratum(dmMesh, topology::Stratum::HEIGHT, 0);
+    const PetscInt cStart = cellsStratum.begin();
+    const PetscInt cEnd = cellsStratum.end();
+
+    const char* const labelName = pylith::topology::Mesh::getCellsLabelName();
+    for (PetscInt c = cStart; c < cEnd; ++c) {
+        err = DMSetLabelValue(dmMesh, labelName, c, 1);PYLITH_CHECK_ERROR(err);
+    } // for
+
+    
+    DMLabel faceLabel;
+    err = DMGetLabel(dmMesh, "Face Sets", &faceLabel);PYLITH_CHECK_ERROR(err);PYLITH_CHECK_ERROR(err);
+    err = DMPlexLabelComplete(dmMesh, faceLabel);PYLITH_CHECK_ERROR(err);
+
+    IS is;
+    DMLabel xnegLabel;
+    char xnegName[] = "boundary_xneg";
+    err = DMCreateLabel(dmMesh, xnegName);PYLITH_CHECK_ERROR(err);
+    err = DMGetLabel(dmMesh, xnegName, &xnegLabel);PYLITH_CHECK_ERROR(err);
+    err = DMLabelGetStratumIS(faceLabel, 1, &is);PYLITH_CHECK_ERROR(err);
+    err = DMLabelSetStratumIS(xnegLabel, 1, is);PYLITH_CHECK_ERROR(err);
+    
+    DMLabel xposLabel;
+    char xposName[] = "boundary_xpos";
+    err = DMCreateLabel(dmMesh, xposName);PYLITH_CHECK_ERROR(err);
+    err = DMGetLabel(dmMesh, xposName, &xposLabel);PYLITH_CHECK_ERROR(err);
+    err = DMLabelGetStratumIS(faceLabel, 3, &is);PYLITH_CHECK_ERROR(err);
+    err = DMLabelSetStratumIS(xposLabel, 1, is);PYLITH_CHECK_ERROR(err);
```

Comment:
@baagaard @knepley What would you like this to look like at the user interface level? Should there be a new schema for `label_name:value` so that one can distinguish different values of the label?

Reply:
We currently have
```toml
label = boundary_xneg
```
@baagaard @knepley What would you think of this?
```toml
label.name = "Face Sets"
label.value = 3
```
It could accept strings (once [string - int tables](https://gitlab.com/petsc/petsc/-/issues/915) are implemented):
```toml
label.name = "Face Sets"
label.value = "North"
```
An alternative could look like
```toml
[label]
"Face Sets" = 3
```

Sometimes one wishes to assign the same boundary condition to several face sets. This could be like
```toml
[label]
"Face Sets" = [3,5,9,22]
```
or
```toml
[label]
blue_faces = [3,4]
green_faces = [11,13]
```
This decomposition isn't directly supported by `DMAddBoundary` and arguably reflects disorganization. Anyway, what sort of schema do you like and consider extensible to the Exodus meshes you know about and the things you can imagine people may want to do with Gmsh and other meshes PETSc can create/read?

Reply:
I think the following is relatively robust and would work in most cases:
```
label.name = Face Sets
label.values = [2, 3, 4]
```

We can set `label.values = [1]` as the default to minimize the changes for current users.

Reply:
Wonderful, thanks! I hope you and your family are feeling better. Does the Pyre parser work this way (`label.value`) like TOML, or should we use underscores like seems to be elsewhere
```toml
label = Face Sets
label_value = [2, 3, 4]
```
Short-term benefit of this is that existing users don't need to change anything.

Reply:
Right now, label is just a string. We will need to change it to an object that has `name` (str) and `value` (list) Pyre properties.

The .cfg syntax will be
````
label.name = Face Sets
label.values = [2, 3, 4]
```